### PR TITLE
8299518: HotSpotVirtualMachine shared code across different platforms

### PR DESCRIPTION
**Pull request by y1yang0**
harmless refactor to share code across different platforms of VirtualMachineImpl:
1. Shared code to process command response after requesting a command execution
2. Read functionality in SocketInputStream can be reused

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8299518](https://bugs.openjdk.org/browse/JDK-8299518): HotSpotVirtualMachine shared code across different platforms

### Reviewers
 * [Chris Plummer](https://openjdk.org/census#cjplummer) (@plummercj - **Reviewer**) ⚠️ Review applies to [6033fcc4](https://git.openjdk.org/jdk/pull/11823/files/6033fcc4b06b342e55e6f95ac7d1b2efd86b1a24)
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**) ⚠️ Review applies to [6033fcc4](https://git.openjdk.org/jdk/pull/11823/files/6033fcc4b06b342e55e6f95ac7d1b2efd86b1a24)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/11823/head:pull/11823` \
`$ git checkout pull/11823`

Update a local copy of the PR: \
`$ git checkout pull/11823` \
`$ git pull https://git.openjdk.org/jdk pull/11823/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 11823`

View PR using the GUI difftool: \
`$ git pr show -t 11823`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/11823.diff">https://git.openjdk.org/jdk/pull/11823.diff</a>

</details>

### CODE REVIEW

**Comment by dholmes-ora** [comment]
There are a lot of stylistic changes here that have nothing to do with refactoring or code sharing and they make the real changes very hard to see. At a minimum things like variable renaming should be done in a separate commit, so the refactoring is more obvious, but it may be better to use a separate cleanup RFE.

File: src/jdk.attach/aix/classes/sun/tools/attach/VirtualMachineImpl.java [hunk 1]
```diff
@@ -57,51 +56,46 @@ public class VirtualMachineImpl extends HotSpotVirtualMachine {
...
         // This provider only understands pids
-        int pid;
-        try {
-            pid = Integer.parseInt(vmid);
-            if (pid < 1) {
-                throw new NumberFormatException();
-            }
-        } catch (NumberFormatException x) {
-            throw new AttachNotSupportedException("Invalid process identifier: " + vmid);
+        int pid = Integer.parseInt(vmid);
```
Comment (dholmes-ora) [inline src/jdk.attach/aix/classes/sun/tools/attach/VirtualMachineImpl.java:59]: This will now throw `NumberFormatException` instead of the expected `AttachNotSupportedException`.

Comment (y1yang0) [reply src/jdk.attach/aix/classes/sun/tools/attach/VirtualMachineImpl.java:59]: The same check is performed in its superclass HotSpotVirtualMachine. So we can be sure that it will not throw NumberFormatException here

https://github.com/openjdk/jdk/blob/38cfc591725de478879266584280562f0ba4b42f/src/jdk.attach/share/classes/sun/tools/attach/HotSpotVirtualMachine.java#L69-L74

Comment (dholmes-ora) [reply src/jdk.attach/aix/classes/sun/tools/attach/VirtualMachineImpl.java:59]: In that case the `int pid` determined by the superclass should be made available to the subclasses so they can use it and not need to re-parse the `vmid`.

File: src/jdk.attach/linux/classes/sun/tools/attach/VirtualMachineImpl.java [hunk 2]
```diff
@@ -237,40 +192,19 @@ InputStream execute(String cmd, Object ... args) throws AgentLoadException, IOEx
     /*
      * InputStream for the socket connection to get target VM
      */
-    private static class SocketInputStream extends InputStream {
-        int s = -1;
-
-        public SocketInputStream(int s) {
-            this.s = s;
+    private static class SocketInputStreamImpl extends SocketInputStream {
```
Comment (dholmes-ora) [inline src/jdk.attach/linux/classes/sun/tools/attach/VirtualMachineImpl.java:195]: Can this class definition also be shared by making it a protected nested class in the superclass?

Comment (dholmes-ora) [reply src/jdk.attach/linux/classes/sun/tools/attach/VirtualMachineImpl.java:195]: Sorry I see it already is, but I think we can do better.

File: src/jdk.attach/share/classes/sun/tools/attach/HotSpotVirtualMachine.java [hunk 3]
```diff
@@ -366,6 +369,87 @@ String readErrorMessage(InputStream in) throws IOException {
...
+     */
+    static abstract class SocketInputStream extends InputStream {
+        private final long fd;
+
+        public SocketInputStream(long fd) {
+            this.fd = fd;
+        }
+
+        protected abstract int readImpl(long fd, byte[] bs, int off, int len) throws IOException;
+        protected abstract void closeImpl(long fd) throws IOException;
```
Comment (dholmes-ora) [inline src/jdk.attach/share/classes/sun/tools/attach/HotSpotVirtualMachine.java:425]: If the subclasses all override these in exactly the same way then these do not need to be abstract and can simply delegate to VirtualMachineImpl.xxx

File: src/jdk.attach/linux/classes/sun/tools/attach/VirtualMachineImpl.java (hunk 2 above)
Comment (y1yang0) [reply src/jdk.attach/linux/classes/sun/tools/attach/VirtualMachineImpl.java:195]: This(`SocketInputStreamImpl`) is implementation-specific, it calls platform-dependent VirtualMachine.xxx.

File: src/jdk.attach/share/classes/sun/tools/attach/HotSpotVirtualMachine.java [hunk 4]
```diff
@@ -183,6 +184,8 @@ public void loadAgent(String agent, String options)
     private static final int ATTACH_ERROR_NOTONCP       = 101;
     private static final int ATTACH_ERROR_STARTFAIL     = 102;
 
+    // known error
+    private static final int ATTACH_ERROR_BADVERSION = 101;
```
[the remark this answers is not in the document]
Comment (y1yang0) [reply src/jdk.attach/share/classes/sun/tools/attach/HotSpotVirtualMachine.java:188]: They are located in different error families. The former is to deal with agent_load and the latter is to deal with command execution version errors.

File: src/jdk.attach/share/classes/sun/tools/attach/HotSpotVirtualMachine.java (hunk 3 above)
Comment (y1yang0) [reply src/jdk.attach/share/classes/sun/tools/attach/HotSpotVirtualMachine.java:425]: All Posix OS platforms call the same VirtualMachineImpl.read/write/etc, but Windows is an unusual guy, it calls VirtualMachineImpl.readPipe/writePipe/etc

**Comment by sspitsyn** [comment]
I like the approach in general.
Also, I agree with David on his comments, especially on the renaming.
The abstract methods `readImpl()` and `closeImpl()` is better to name as `read()` and `close()`.

File: src/jdk.attach/aix/classes/sun/tools/attach/VirtualMachineImpl.java (hunk 1 above)
Comment (dholmes-ora) [reply src/jdk.attach/aix/classes/sun/tools/attach/VirtualMachineImpl.java:59]: It is well within scope of improving code sharing in this arera, but I'm okay with leaving that for a next step.

File: src/jdk.attach/share/classes/sun/tools/attach/HotSpotVirtualMachine.java (hunk 3 above)
Comment (dholmes-ora) [reply src/jdk.attach/share/classes/sun/tools/attach/HotSpotVirtualMachine.java:425]: That is a shame, though perhaps we could just rename those methods on Windows?

This also raises the obvious question can we in fact reduce this to just a Windows and Posix version, or are there other differences between Linux, macOS etc that have to be accounted for? Though in that case we could introduce a shared superclass for the Posix platforms. Next steps perhaps?

Comment (y1yang0) [reply src/jdk.attach/share/classes/sun/tools/attach/HotSpotVirtualMachine.java]: The remaining code for the POSIX system is exactly the same, except that Linux will make an extra attempt to find java_pid in cgroup environment.

File: src/jdk.attach/share/classes/sun/tools/attach/HotSpotVirtualMachine.java [hunk 5]
```diff
@@ -366,6 +369,87 @@ String readErrorMessage(InputStream in) throws IOException {
         return message.toString();
     }
 
+    void processCompletionStatus(IOException ioe, String cmd, InputStream sis) throws AgentLoadException, IOException {
```
[the remark this answers is not in the document]
Comment (dholmes-ora) [reply src/jdk.attach/share/classes/sun/tools/attach/HotSpotVirtualMachine.java:375]: I was thinking more about describing what the parameters are - in particular it is unclear why the actual IOE can be replaced by a passed in one. If we wanted a custom message then I would at least expect to chain the actual IOE to the replacement.

File: src/jdk.attach/windows/classes/sun/tools/attach/VirtualMachineImpl.java [hunk 6]
```diff
@@ -126,12 +126,12 @@ public SocketInputStreamImpl(long fd) {
         }
 
         @Override
-        protected int readImpl(long fd, byte[] bs, int off, int len) throws IOException {
+        protected int read(long fd, byte[] bs, int off, int len) throws IOException {
             return VirtualMachineImpl.readPipe(fd, bs, off, len);
```
Comment (dholmes-ora) [inline src/jdk.attach/windows/classes/sun/tools/attach/VirtualMachineImpl.java:130]: If you add read/close methods for Windows that then call readPipe/closePipe, then we don't need to subclass and override.

Comment (dholmes-ora) [reply src/jdk.attach/windows/classes/sun/tools/attach/VirtualMachineImpl.java:130]: Looks like Windows needs a long fd so the shared API takes long and we then cast to int for the Unix native methods. A little messy but fixable.

Comment (y1yang0) [reply src/jdk.attach/windows/classes/sun/tools/attach/VirtualMachineImpl.java:130]: > If you add read/close methods for Windows that then call readPipe/closePipe, then we don't need to subclass and override.

Sorry, I don't understand what this means. VirtualMachineImpl is platform-dependent while HotSpotVirtualMachine is shared, we always need to sub-class socket input stream in order to call platform-depend read/close.

Comment (dholmes-ora) [reply src/jdk.attach/windows/classes/sun/tools/attach/VirtualMachineImpl.java:130]: As long as all the platform-specific `VirtualMachineImpl` classes define a `read` and `close` method then we can simply have the shared socket stream class call those methods. Windows will need an extra level of indirection because its `read` method will have to call `readPipe`.

Comment (y1yang0) [reply src/jdk.attach/windows/classes/sun/tools/attach/VirtualMachineImpl.java:130]: > As long as all the platform-specific `VirtualMachineImpl` classes define a `read` and `close` method then we can simply have the shared socket stream class call those methods. Windows will need an extra level of indirection because its `read` method will have to call `readPipe`.

```
/path/src/jdk.attach/share/classes/sun/tools/attach/HotSpotVirtualMachine.java:444: error: incompatible types: possible lossy conversion from long to int
            return VirtualMachineImpl.read(fd, bs, off, len);
                                           ^
/path/src/jdk.attach/share/classes/sun/tools/attach/HotSpotVirtualMachine.java:449: error: incompatible types: possible lossy conversion from long to int
                VirtualMachineImpl.close(fd);
```
Unfortunately, this causes compilation error because we have `long fd` in shard socket stream and  `read(int fd...)/readPipe(long fd)` in VirtualMachineImpl

**Comment by plummercj** [comment]
I would also like to request that the variable renames be omitted from this PR. They create too much noise and would best left to a PR that focuses on just the renames and possibly other style changes.

File: src/jdk.attach/share/classes/sun/tools/attach/HotSpotVirtualMachine.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (y1yang0) [reply src/jdk.attach/share/classes/sun/tools/attach/HotSpotVirtualMachine.java:377]: > I was thinking more about describing what the parameters are - in particular it is unclear why the actual IOE can be replaced by a passed in one. If we wanted a custom message then I would at least expect to chain the actual IOE to the replacement.

Done. Parameter ioe means if we get IOE during previous command execution, delay throwing it until command execution completion status has been read. Since all other parameters are straightforward and this is a private method, I don't write  strandard `@param` descriptions for them.

**Comment by y1yang0** [comment]
> I would also like to request that the variable renames be omitted from this PR. They create too much noise and would best left to a PR that focuses on just the renames and possibly other style changes.

Okay, reverted.

File: src/jdk.attach/windows/classes/sun/tools/attach/VirtualMachineImpl.java (hunk 6 above)
Comment (dholmes-ora) [reply src/jdk.attach/windows/classes/sun/tools/attach/VirtualMachineImpl.java]: That is unfortunate.

File: src/jdk.attach/share/classes/sun/tools/attach/HotSpotVirtualMachine.java [hunk 7]
```diff
@@ -366,6 +369,92 @@ String readErrorMessage(InputStream in) throws IOException {
...
+                }
+                throw new AttachOperationFailedException(message);
+            }
+        }
+    }
+
+    /*
+     * InputStream for the socket connection to get target VM
+     */
+    static abstract class SocketInputStream extends InputStream {
```
Comment (turbanoff) [inline src/jdk.attach/share/classes/sun/tools/attach/HotSpotVirtualMachine.java:422]: Use blessed modifiers order
```suggestion
    abstract static class SocketInputStream extends InputStream {
```

File: src/jdk.attach/aix/classes/sun/tools/attach/VirtualMachineImpl.java [hunk 8]
```diff
@@ -57,14 +56,9 @@ public class VirtualMachineImpl extends HotSpotVirtualMachine {
...
+        if (pid < 1) {
+            throw new AttachNotSupportedException("Invalid process identifier -1");
```
Comment (plummercj) [inline src/jdk.attach/aix/classes/sun/tools/attach/VirtualMachineImpl.java:61]: Why the change to using -1 instead of `vmid`? It seems presumptuous that it is -1, and also seems less useful than printing `vmid`.

File: src/jdk.attach/share/classes/sun/tools/attach/HotSpotVirtualMachine.java [hunk 9]
```diff
@@ -366,6 +369,92 @@ String readErrorMessage(InputStream in) throws IOException {
...
+            } else if (len == 0) {
+                return 0;
+            }
+            return read(fd, bs, off, len);
+        }
+
+        public synchronized void close() throws IOException {
+            if (fd != -1) {
+                close(fd);
+            }
```
Comment (plummercj) [inline src/jdk.attach/share/classes/sun/tools/attach/HotSpotVirtualMachine.java:455]: There used to be logic to set `fd` (previously called `s`) to -1 during the close operation. Is that no longer needed?

Comment (y1yang0) [reply src/jdk.attach/share/classes/sun/tools/attach/HotSpotVirtualMachine.java:455]: I don't see why we need it because this method is protected by synchronized.

**Comment by y1yang0** [comment]
Can I have any review? Thanks.

File: src/jdk.attach/share/classes/sun/tools/attach/HotSpotVirtualMachine.java (hunk 9 above)
Comment (plummercj) [reply src/jdk.attach/share/classes/sun/tools/attach/HotSpotVirtualMachine.java:455]: You will end up leaving fd set even after the close completes.  A subsequent close (which I assume should never happen) would end up with an IOException. Previously close would end up not doing anything when called again, and simply return with no error. I'm not sure which is the best behavior in this case, but it is a change in behavior.

**Review by plummercj — APPROVED** [review]

**Comment by y1yang0** [comment]
@sspitsyn @dholmes-ora @turbanoff May I ask your help to review this patch? Thanks.

File: src/jdk.attach/share/classes/sun/tools/attach/HotSpotVirtualMachine.java [hunk 10]
```diff
@@ -366,6 +369,94 @@ String readErrorMessage(InputStream in) throws IOException {
         return message.toString();
     }
 
+    /*
+     * Utility method to process the completion status after command execution.
+     * If we get IOE during previous command execution, delay throwing it until
+     * completion status have been read.
```
Comment (dholmes-ora) [inline src/jdk.attach/share/classes/sun/tools/attach/HotSpotVirtualMachine.java:375]: Nit: s/have/has/

**Review by dholmes-ora — APPROVED** [review]
Nothing further from me.

Thanks.
